diff -up cyrus-sasl-2.1.23/lib/Makefile.am.relro cyrus-sasl-2.1.23/lib/Makefile.am
--- cyrus-sasl-2.1.23/lib/Makefile.am.relro	2009-04-28 17:09:15.000000000 +0200
+++ cyrus-sasl-2.1.23/lib/Makefile.am	2011-08-13 22:35:16.482216782 +0200
@@ -60,7 +60,7 @@ LIB_DOOR= @LIB_DOOR@
 lib_LTLIBRARIES = libsasl2.la
 
 libsasl2_la_SOURCES = $(common_sources) $(common_headers)
-libsasl2_la_LDFLAGS = -version-info $(sasl_version)
+libsasl2_la_LDFLAGS = -Wl,-z,relro -version-info $(sasl_version)
 libsasl2_la_DEPENDENCIES = $(LTLIBOBJS)
 libsasl2_la_LIBADD = $(LTLIBOBJS) $(SASL_DL_LIB) $(LIB_SOCKET) $(LIB_DOOR)
 
diff -up cyrus-sasl-2.1.23/plugins/Makefile.am.relro cyrus-sasl-2.1.23/plugins/Makefile.am
--- cyrus-sasl-2.1.23/plugins/Makefile.am.relro	2011-08-13 22:35:16.157216782 +0200
+++ cyrus-sasl-2.1.23/plugins/Makefile.am	2011-08-13 22:35:16.497091719 +0200
@@ -62,7 +62,7 @@ ldapdb_version = 2:23:0
 srp_version = 2:23:0
 
 INCLUDES=-I$(top_srcdir)/include -I$(top_srcdir)/lib -I$(top_srcdir)/sasldb -I$(top_builddir)/include
-AM_LDFLAGS = -module -export-dynamic -rpath $(plugindir)
+AM_LDFLAGS = -Wl,-z,relro -module -export-dynamic -rpath $(plugindir)
 
 COMPAT_OBJS = @LTGETADDRINFOOBJS@ @LTGETNAMEINFOOBJS@ @LTSNPRINTFOBJS@
 
diff -up cyrus-sasl-2.1.23/saslauthd/Makefile.am.relro cyrus-sasl-2.1.23/saslauthd/Makefile.am
--- cyrus-sasl-2.1.23/saslauthd/Makefile.am.relro	2011-08-13 22:39:09.241091660 +0200
+++ cyrus-sasl-2.1.23/saslauthd/Makefile.am	2011-08-13 22:48:36.727091540 +0200
@@ -14,6 +14,8 @@ saslauthd_SOURCES = mechanisms.c globals
 		    md5.c saslauthd_md5.h md5global.h 
 EXTRA_saslauthd_sources = getaddrinfo.c getnameinfo.c
 saslauthd_DEPENDENCIES = saslauthd-main.o @LTLIBOBJS@
+saslauthd_CFLAGS = -fPIE
+saslauthd_LDFLAGS = -pie -Wl,-z,relro -Wl,-z,now
 saslauthd_LDADD	= @SASL_KRB_LIB@ \
 		  @GSSAPIBASE_LIBS@ @GSSAPI_LIBS@ @LIB_CRYPT@ @LIB_SIA@ \
 		  @LIB_SOCKET@ @SASL_DB_LIB@ @LIB_PAM@ @LDAP_LIBS@ @LTLIBOBJS@
